### ISSUE TITLE
best_iter and best_score problem without specifying early_stopping_rounds

### ISSUE THREAD

**Issue by alastairrushworth** [Microsoft/LightGBM#1353]
Firstly, thanks for a fabulous piece of software!  

There seems to be an issue that `lgb.cv` doesn't print correctly the result of the cross validation if `early_stopping_rounds` is not specified in the params.  Perhaps that is the expected behaviour - but in that case, I'd point out that the documentation for the R package doesn't seem to reflect this (see for example [https://github.com/Microsoft/LightGBM/blob/master/R-package/demo/cross_validation.R](https://github.com/Microsoft/LightGBM/blob/master/R-package/demo/cross_validation.R)).

Here is the example from the docs:

```
require(lightgbm)
# load in the agaricus dataset
data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")
dtrain <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label)
dtest <- lgb.Dataset(agaricus.test$data, label = agaricus.test$label)

nrounds <- 2
param <- list(num_leaves = 4,
              learning_rate = 1,
              objective = "binary")

print("Running cross validation")
# Do cross validation, this will print result out as
# [iteration]  metric_name:mean_value+std_value
# std_value is standard deviation of the metric
lgb.cv(param,
       dtrain,
       nrounds,
       nfold = 5,
       eval = "binary_error") 
```

The above returns the following output:
```
<lgb.CVBooster>
  Public:
    best_iter: -1
    best_score: -1
    boosters: list
    initialize: function (x) 
    record_evals: list
    reset_parameter: function (new_params) 
```

And with the `early_stopping_rounds` specified:
```
require(lightgbm)
# load in the agaricus dataset
data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")
dtrain <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label)
dtest <- lgb.Dataset(agaricus.test$data, label = agaricus.test$label)

nrounds <- 2
param <- list(num_leaves = 4,
              learning_rate = 1,
              objective = "binary", 
              early_stopping_rounds = 50)

print("Running cross validation")
# Do cross validation, this will print result out as
# [iteration]  metric_name:mean_value+std_value
# std_value is standard deviation of the metric
lgb.cv(param,
       dtrain,
       nrounds,
       nfold = 5,
       eval = "binary_error")
```
The result is 

```
<lgb.CVBooster>
  Public:
    best_iter: 2
    best_score: -0.0222631691252492
    boosters: list
    initialize: function (x) 
    record_evals: list
    reset_parameter: function (new_params)  
```

**Comment by guolinke** [comment]
I think it is by design. when without early_stopping, the best_iter and best_score will never be set. 
best_iter=-1 means using the last iteration.  

@Laurae2 It is changed a while ago, due to some feedbacks, but i don't remember why.
Are there any problems ?

**Comment by Laurae2** [comment]
@guolinke For time series (or any data including a treatment factor), it is important to check the evolution of the metric by treatment/time over the number of iterations. Many times, one treatment/time will be more noisy than others, and could trigger the early stopping when it should not.

**Comment by guolinke** [comment]
@Laurae2 I think the trigger condition is the same as before. 
The only difference is, the best_iter is set to the best iteration, that is `stop_iteration - es_rounds`.

**Comment by guolinke** [comment]
close now. feel free to reopen if still have questions.
